### PR TITLE
Archives: Properly implemented the SystemSaveData archive.

### PR DESCRIPTION
**Pull request by Subv**
Required to get Home Menu further.

### CODE REVIEW

File: src/core/hle/service/cfg/cfg.cpp [hunk 1]
```diff
@@ -158,27 +158,28 @@ ResultCode FormatConfig() {
...
-    if (file != nullptr) {
-        file->Read(0, CONFIG_SAVEFILE_SIZE, cfg_config_file_buffer.data());
+    ASSERT_MSG(archive_result.Succeeded(), "Could not open the CFG SystemSaveData archive!");
+    cfg_system_save_data_archive = *archive_result;
+
+    FileSys::Path config_path("config");
+    FileSys::Mode open_mode = {};
+    open_mode.read_flag = 1;
+    auto config_result = Service::FS::OpenFileFromArchive(*archive_result, config_path, open_mode);
+    // Read the file if it already exists
```
Comment (bunnei) [inline src/core/hle/service/cfg/cfg.cpp]: could use a space above this comment

File: src/core/hle/service/cfg/cfg.cpp [hunk 2]
```diff
@@ -158,27 +158,28 @@ ResultCode FormatConfig() {
...
-    std::string nand_directory = FileUtil::GetUserPath(D_NAND_IDX);
-    cfg_system_save_data = Common::make_unique<FileSys::Archive_SystemSaveData>(
-                           nand_directory, CFG_SAVE_ID);
-    if (!cfg_system_save_data->Initialize()) {
-        LOG_CRITICAL(Service_CFG, "Could not initialize SystemSaveData archive for the CFG:U service");
-        return;
+    // Open the SystemSaveData archive 0x00010017
+    FileSys::Path archive_path(cfg_system_savedata_id);
+    auto archive_result = Service::FS::OpenArchive(Service::FS::ArchiveIdCode::SystemSaveData, archive_path);
+    // If the archive didn't exist, create the files inside
```
Comment (bunnei) [inline src/core/hle/service/cfg/cfg.cpp]: could use a space above this comment

File: src/core/hle/service/cfg/cfg.cpp [hunk 3]
```diff
@@ -107,9 +106,10 @@ ResultCode UpdateConfigNANDSavegame() {
...
     mode.create_flag = 1;
     FileSys::Path path("config");
-    auto file = cfg_system_save_data->OpenFile(path, mode);
-    ASSERT_MSG(file != nullptr, "could not open file");
-    file->Write(0, CONFIG_SAVEFILE_SIZE, 1, cfg_config_file_buffer.data());
+    auto config_result = Service::FS::OpenFileFromArchive(cfg_system_save_data_archive, path, mode);
+    ASSERT_MSG(config_result.Succeeded(), "could not open file");
+    auto config = config_result.MoveFrom();
+    config->backend->Write(0, CONFIG_SAVEFILE_SIZE, 1, cfg_config_file_buffer.data());
     return RESULT_SUCCESS;
```
Comment (bunnei) [inline src/core/hle/service/cfg/cfg.cpp]: could use some more white space in this function

**Comment by bunnei** [comment]
LGTM. Only real complaint is for some reason there are several places in code where I would prefer you had some new lines to make it more readable.

**Comment by Subv** [comment]
Updated

File: src/core/file_sys/archive_systemsavedata.h [hunk 4]
```diff
@@ -15,17 +15,17 @@
...
-     * @return true if it initialized successfully
-     */
-    bool Initialize();
+    ResultVal<std::unique_ptr<ArchiveBackend>> Open(const Path& path) override;
+    ResultCode Format(const Path& path) override;
 
     std::string GetName() const override { return "SystemSaveData"; }
+
+private:
+    std::string mount_point;
```
Comment (yuriks) [inline src/core/file_sys/archive_systemsavedata.h]: I'd prefer if you named this `base_path` or similar.

**Comment by yuriks** [comment]
Mostly looks good.

**Comment by linkmauve** [comment]
LGTM

**Comment by Subv** [comment]
Updated and rebased

**Comment by yuriks** [comment]
LGTM

**Pull request merged**